core/src/main/java/com/linecorp/armeria/common/util/UnmodifiableFuture.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.util;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.internal.eventloop.EventLoopCheckingCompletableFuture;
+
+/**
+ * A {@link CompletableFuture} which prevents the caller from completing it. An attempt to call any of
+ * the following methods will trigger an {@link UnsupportedOperationException}:
+ * <ul>
+ *   <li>{@link #complete(Object)}</li>
+ *   <li>{@link #completeExceptionally(Throwable)}</li>
+ *   <li>{@link #obtrudeValue(Object)}</li>
+ *   <li>{@link #obtrudeException(Throwable)}</li>
+ * </ul>
+ * Also, {@link #cancel(boolean)} will do nothing but returning whether cancelled or not.
+ */
+public final class UnmodifiableFuture<T> extends EventLoopCheckingCompletableFuture<T> {
+
+    private static final UnmodifiableFuture<?> NIL;
+
+    static {
+        NIL = new UnmodifiableFuture<>();
+        NIL.doComplete(null);
+    }
+
+    /**
+     * Returns an {@link UnmodifiableFuture} which has been completed with the specified {@code value}.
+     */
+    public static <U> UnmodifiableFuture<U> completedFuture(@Nullable U value) {
```

Comment:
Shouldn't this return `CompletableFuture`?

Reply:
On second thought, it's better to return `UnmodifiableFuture`. 😉 Sorry about the noise.

Reply:
I'm not sure. It's like `ImmutableList.of()` returns an `ImmutableList` instead of a `List`. We might add more public API to it later as well.